### PR TITLE
tui: handle Smulx extension capability (extended underline)

### PR DESCRIPTION
**Pull request by blueyed**
This will be added to terminfo for vte and tmux.

Ref: https://github.com/tmux/tmux/issues/1492#issuecomment-427675180

### CODE REVIEW

File: src/nvim/tui/tui.c [hunk 1]
```diff
@@ -531,8 +532,9 @@ static void update_attrs(UI *ui, HlAttrs attrs)
   if (italic) {
     unibi_out(ui, unibi_enter_italics_mode);
   }
-  if (undercurl && data->unibi_ext.enter_undercurl_mode) {
-    unibi_out_ext(ui, data->unibi_ext.enter_undercurl_mode);
+  if (undercurl && data->unibi_ext.extended_underline_mode != -1) {
```
Comment (blueyed) [inline src/nvim/tui/tui.c]: Not checking for `-1` is a problem currently already (also above).

**Comment by justinmk** [comment]
unibi_ext.extended_underline_mode

it's already in `unibi_ext` so "extended" is implied. Perhaps `unibi_ext.set_underline_style` ?

I notice it's not an enter/exit pair (cf. `{enter,exit}_undercurl_mode`)? How does that work?

**Comment by blueyed** [comment]
> unibi_ext.set_underline_style

:+1: 

As mentioned before the exit codes were never used.  I assume it works because it gets reset when setting "everything"?  Worth investigating though.

**Comment by blueyed** [comment]
@justinmk 
Rebased it for now, but still consider it WIP.
I also have to patch it locally to detect/handle tmux in kitty also.
It works with gnome-terminal, but only because/if `$VTE_VERSION` is set in its env.

**Comment by blueyed** [comment]
I think this is ready.
Just tried it using ncurses master and `tmux -L new -f tmux-test.conf` in gnome-terminal, with `unset VTE_VERSION` and `set -as terminal-overrides ',*:Smulx=\E[4::%p1%dm'` in `tmux-test.conf`:

`TERM=tmux TERMINFO=…/ncurses/build/inst/share/terminfo build/bin/nvim`

`:hi SpellBad` then was undercurl with this PR (814fc4e), but not on master.

**Comment by petobens** [comment]
Hi @blueyed!  I'm trying to get undercurl to work with neovim (latest master) under tmux (2.9a) while using kitty (0.14.6) as my
terminal. I have:

i) added the following to my tmux config:
```
set-option -s -a terminal-overrides ',xterm*:sitm=\E[3m,*:Smulx=\E[4::%p1%dm'
```
so that running `tmux info | grep Sm` reports `199: Smulx: (string) \033[4:%p1%dm`

ii) added an entry to my custom terminfo so that it looks like this:
```
tmux-256color|tmux-xterm with 256 colors and italics,
    use=xterm-256color,
    sitm=\E[3m,
    ritm=\E[23m,
    Smulx=\E[4:%p1%dm,
```
and installed the above with `tic -x`.

However I cannot seem to get undercurl to work under tmux and neovim. With the following
minimal `init.vim` file:

```viml
filetype plugin indent on
set termguicolors
set spell

hi SpellBad gui=undercurl guisp=#e06c75

" A mispeled word
```
In the following image I have tried this both with (left image) and without tmux (right
image). It can be seen that undercurl generally works except for the combination
tmux-nvim. Any pointers are greatly appreciated. Thanks!

Screenshot_2019-11-11_16:56:27

**Comment by petobens** [comment]
Hi @blueyed sorry to bother you again. Any ideas why I cannot seem to get colored undercurl support in neovim? Just for the record (following my comment above) in tmux (3.0a) everything works as expected:

image

**Comment by petobens** [comment]
@justinmk @blueyed should I open an issue with my comments above? Thanks!

**Comment by justinmk** [comment]
I don't want to debug configs. Please narrow down the issue and send a specific suggestion or patch.

**Comment by petobens** [comment]
Seems to be working now  (no idea why though since I didn't touch the configs). If I ever find out what's going on i'll report it here. Thanks all for the help

**Comment by petobens** [comment]
Finally figured it out: I had a custom terminfo also called `tmux-256colors`. Removing that and using the one provided by `ncurses` makes things start to work.

**Comment by kabouzeid** [comment]
@petobens could you provide more details on how you fixed this? I have the same exact problem you described.

**Comment by akinsho** [comment]
@kabouzeid I managed to get this working in `tmux` with `kitty` term using
```tmux
# Allow undercurls for terminals that support them.
set -as terminal-overrides ',*:Smulx=\E[4::%p1%dm'
# Allow coloured undercurls for terminals that support them.
set -as terminal-overrides ',*:Setulc=\E[58::2::%p1%{65536}%/%d::%p1%{256}%/%{255}%&%d::%p1%{255}%&%d%;m'
```
I didn't touch anything else

**Comment by kabouzeid** [comment]
So weird, I have the same exact thing in my `.tmux.conf`.

But for me undercurl:
- works in kitty
- works in kitty -> neovim
- works in kitty -> tmux
- does *not* work in kitty -> tmux -> neovim

What is the output of `echo $TERM` and `echo $TERMINFO` for you when you are in tmux?

**Comment by kabouzeid** [comment]
Also copied `set -g default-terminal "${TERM}"` from your config, and now I'm at least getting a colored underline in kitty->tmux->neovim. But still not curly.

**Comment by akinsho** [comment]
```
TERMINFO=/Applications/kitty.app/Contents/Resources/kitty/terminfo
TERM=xterm-kitty
```
I set my tmux term using `set -g default-terminal "${TERM}"` I found this whilst digging through issues in kitty's repo

**Comment by kabouzeid** [comment]
Yeah, thanks anyways :D I can live with the colored but non curly underlines for now, I'll look into this again later.

**Comment by kabouzeid** [comment]
Ok I've found it. Turns out the colorscheme I'm using (sonokai) disables undercurl for tmux https://github.com/sainnhe/sonokai/blob/764bd716f08f1441e0020b8ae0c8d1b53970e4a9/autoload/sonokai.vim#L163

**Pull request merged**